### PR TITLE
Define pointer arith

### PR DESCRIPTION
**Pull request by fbs**
##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: src/ast/semantic_analyser.cpp [hunk 1]
```diff
@@ -1594,6 +1594,68 @@ void SemanticAnalyser::binop_int(Binop &binop)
...
+    LOG(ERROR, binop.loc, err_)
+        << "The " << opstr(binop)
+        << " operator can not be used on expressions of types " << lht << ", "
+        << rht;
+  };
+
+  // Binop on two pointers
+  if (other.IsPtrTy())
+  {
+    if (compare)
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp:1632]: Clang emits a warning, we should probably copy it

test.c:6:8: warning: comparison of distinct pointer types ('char *' and 'int *')
      [-Wcompare-distinct-pointer-types]

File: tests/runtime/pointers [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+NAME u32 pointer increment
+RUN bpftrace -e 'BEGIN { @=(int32*) 32; @+=1; exit(); }'
+EXPECT ^@: 36
+TIMEOUT 5
+
+NAME u64 pointer increment
+RUN bpftrace -e 'BEGIN { @=(int64*) 32; @+=1; exit(); }'
+EXPECT ^@: 40
+TIMEOUT 5
```
Comment (viktormalik) [inline tests/runtime/pointers:104]: Aren't these the same tests as the first 4 in the file?

**Review by viktormalik — COMMENTED** [review]
just some small nits, LGTM overall

File: tests/runtime/pointers (hunk 2 above)
Comment (fbs) [reply tests/runtime/pointers:104]: hmm think i wanted to modify them, but can't remember into what. Good catch :)

**Comment by fbs** [comment]
need to add some codegen tests, pretty sure this has some issues.

File: tests/runtime/pointers (hunk 2 above)
Comment (viktormalik) [reply tests/runtime/pointers]: Perhaps `s/increment/decrement`? ;-)

File: tests/codegen/pointer_inc_var.cpp [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#include "common.h"
+
+namespace bpftrace {
+namespace test {
+namespace codegen {
+
+TEST(codegen, pointer_inc_var)
+{
+  test("kprobe:f { $v = (int16*)1000; $v++ }", NAME);
```
Comment (viktormalik) [inline tests/codegen/pointer_inc_var.cpp:9]: What about doing something like `$x = $v++` to test that the original value of `$v` is assigned? Also, a similar runtime test would be nice.

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,8 @@ and this project adheres to
 #### Added
 - Warn if attaching a kprobe to a non-traceable function
   - [#1835](https://github.com/iovisor/bpftrace/pull/1835)
+- Add C like pointer arithmetic
```
Comment (fbs) [inline CHANGELOG.md:14]: wrong section due to 0.13 release, needs rebase

File: tests/codegen/pointer_inc_var.cpp (hunk 3 above)
Comment (fbs) [reply tests/codegen/pointer_inc_var.cpp:9]: ah yeah won't hurt to have

File: tests/semantic_analyser.cpp [hunk 5]
```diff
@@ -2067,6 +2067,64 @@ TEST(semantic_analyser, double_pointer_struct)
...
+{
+  test(R"_(BEGIN { $t = (int32*) 32; $t = $t + 1 })_", 0);
+  test(R"_(BEGIN { $t = (int32*) 32; $t +=1 })_", 0);
+  // test(R"_(BEGIN { $t = (int32*) 32; $t++ })_", 0);
+  // test(R"_(BEGIN { $t = (int32*) 32; ++$t })_", 0);
+  test(R"_(BEGIN { $t = (int32*) 32; $t = $t - 1 })_", 0);
+  test(R"_(BEGIN { $t = (int32*) 32; $t -=1 })_", 0);
+  // test(R"_(BEGIN { $t = (int32*) 32; $t-- })_", 0);
+  // test(R"_(BEGIN { $t = (int32*) 32; --$t })_", 0);
+
```
Comment (danobi) [inline tests/semantic_analyser.cpp:2080]: Missing a ptr + ptr test?

File: src/ast/codegen_llvm.cpp [hunk 6]
```diff
@@ -1420,114 +1420,95 @@ static bool unop_skip_accept(Unop &unop)
   return false;
 }
 
-void CodegenLLVM::visit(Unop &unop)
+void CodegenLLVM::unop_int(Unop &unop,
+                           ScopedExprDeleter &scoped_del
+                           __attribute__((unused)))
```
Comment (danobi) [inline src/ast/codegen_llvm.cpp:1425]: Why is `scoped_del` being passed but not used?

File: src/ast/codegen_llvm.cpp [hunk 7]
```diff
@@ -1420,114 +1420,95 @@ static bool unop_skip_accept(Unop &unop)
...
-        else
-        {
-          LOG(FATAL) << "invalid expression passed to " << opstr(unop);
-        }
-        break;
-      }
-      case bpftrace::Parser::token::MUL:
+void CodegenLLVM::unop_ptr(Unop &unop,
+                           ScopedExprDeleter &scoped_del
+                           __attribute__((unused)))
```
Comment (danobi) [inline src/ast/codegen_llvm.cpp:1469]: Same as above

**Comment by fbs** [comment]
Thanks for all the feedback. I added a  'complex' runtime test that should be close to a real use case for pointers

File: src/ast/codegen_llvm.cpp [hunk 8]
```diff
@@ -1294,6 +1294,87 @@ void CodegenLLVM::binop_int(Binop &binop)
...
+      break;
+    case bpftrace::Parser::token::LEFT:
+    case bpftrace::Parser::token::RIGHT:
+    case bpftrace::Parser::token::MOD:
+    case bpftrace::Parser::token::BAND:
+    case bpftrace::Parser::token::BOR:
+    case bpftrace::Parser::token::BXOR:
+    case bpftrace::Parser::token::MUL:
+    case bpftrace::Parser::token::DIV:
+      break;
```
Comment (danobi) [inline src/ast/codegen_llvm.cpp:1321]: maybe `assert(false)` here? Might help catch bugs early if anything changes in future

**Review by danobi — APPROVED** [review]
Nice!

**Review by viktormalik — APPROVED** [review]

**Pull request merged**